File: 096 Unique Binary Search Trees.py
'''
Given n, how many structurally unique BST's (binary search trees) that store values 1...n?

For example,
Given n = 3, there are a total of 5 unique BST's.

   1         3     3      2      1
    \       /     /      / \      \
     3     2     1      1   3      2
    /     /       \                 \
   2     1         2                 3
'''

class Solution(object):
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [1 for __ in range(n + 1)]
        for i in range(2, n + 1):
            s = 0
            for j in range(i):
                s += dp[j] * dp[i - 1 - j]
            dp[i] = s
        return dp[-1]


if __name__ == "__main__":
    assert Solution().numTrees(5) == 42

File: java/_096UniqueBinarySearchTrees.java
/**
 * Given n, how many structurally unique BST's (binary search trees) that store values 1...n?
 * <p>
 * For example,
 * Given n = 3, there are a total of 5 unique BST's.
 * <p>
 * 1         3     3      2      1
 *  \       /     /      / \      \
 *   3     2     1      1   3      2
 *  /     /       \                 \
 * 2     1         2                 3
 * <p>
 * Created by drfish on 27/05/2017.
 */
public class _096UniqueBinarySearchTrees {
    public int numTrees(int n) {
        int[] counts = new int[n + 1];
        counts[0] = 1;
        for (int i = 1; i <= n; i++) {
            int count = 0;
            for (int j = 0; j < i; j++) {
                count += counts[j] * counts[i - 1 - j];
            }
            counts[i] = count;
        }
        return counts[n];
    }

    public static void main(String[] args) {
        _096UniqueBinarySearchTrees uniqueBinarySearchTrees = new _096UniqueBinarySearchTrees();
        assert uniqueBinarySearchTrees.numTrees(3) == 5;
        assert uniqueBinarySearchTrees.numTrees(6) == 132;
    }
}
